diff --git a/wandb_promise/promise.py b/wandb_promise/promise.py
index ef9a45f..5a8a7b3 100644
--- a/wandb_promise/promise.py
+++ b/wandb_promise/promise.py
@@ -5,7 +5,6 @@ from threading import RLock
 from types import TracebackType
 from weakref import WeakKeyDictionary

-from six import reraise  # type: ignore
 from .async_ import Async
 from .compat import (
     Future,
@@ -223,7 +222,7 @@ class Promise(Generic[T]):
         elif self._state == STATE_REJECTED:
             if _raise:
                 raise_val = self._fulfillment_handler0
-                reraise(type(raise_val), raise_val, self._traceback)
+                raise raise_val.with_traceback(self._traceback)
             return self._fulfillment_handler0

     def _fulfill(self, value):
